docs/src/manual/nlp.md

```diff
@@ -435,6 +431,41 @@ register(model, :my_square, 1, f, ∇f, ∇²f)
 @NLobjective(model, Min, my_square(x))
 ```
 
+#### Multivariate functions
+
+For multivariate functions, the hessian function `∇²f` must take an
+`AbstractMatrix` as the first argument, the lower-triangular of which is filled
+in-place:
+```@example
+using JuMP #hide
+f(x...) = (1 - x[1])^2 + 100 * (x[2] - x[1]^2)^2
+function ∇f(g, x...)
+    g[1] = 400 * x[1]^3 - 400 * x[1] * x[2] + 2 * x[1] - 2
+    g[2] = 200 * (x[2] - x[1]^2)
+    return
+end
+function ∇²f(H, x...)
+    H[1, 1] = 1200 * x[1]^2 - 400 * x[2] + 2
+    # H[1, 2] = -400 * x[1]  <-- Not needed. Fill the lower-triangular only.
+    H[2, 1] = -400 * x[1]
+    H[2, 2] = 200.0
+    return
+end
+
+model = Model()
+register(model, :rosenbrock, 2, f, ∇f, ∇²f)
+@variable(model, x[1:2])
+@NLobjective(model, Min, rosenbrock(x[1], x[2]))
+```
+
+!!! warning
+    Make sure the first argument to `∇²f` supports an `AbstractMatrix`, and do
+    not assume the input is `Float64`. You may assume the matrix is initialized
+    with zeros, so you need only to fill in the non-zero terms. The matrix type
+    passed in as `H` depends on the automatic differentiation system, so it may
+    be something other than `Matrix{Float64}`. You may assume only that it
+    supports `size(H)` and `setindex!`.
+
```

Comment:
Add a warning that the hessian matrix is treated as dense, so the performance will be poor on functions with high-dimensional input.